### PR TITLE
perf(pre-push): scope the Go coverage run to the packages you changed

### PR DESCRIPTION
**Pull request by hugalafutro**
## The number

Same commit, same gate, same verdict, measured end to end by feeding the hook the ref line git feeds it:

| | time | what it ran |
|---|---|---|
| before | **198s** | `go test ./internal/...` |
| after | **35s** | `go test ./internal/db ./internal/quota` |

The test change was deliberately realistic: an error string edited inside a function that is actually called in `internal/db`, plus a leaf change in `internal/quota`. Almost the entire 198s was `internal/api`, a package the change never went near.

## Why the whole tree was never needed

`go test -coverprofile` credits coverage **only to the package under test** unless `-coverpkg` widens it, and this repo does not use it. A profile built from one package contains that package's statements and nothing else. Since the gate measures *changed lines*, a dependent package's tests contribute exactly nothing to the number `diff_coverage.py` computes.

Verified rather than assumed, by measuring the same packages both inside the full tree and alone:

| package | in `./internal/...` | alone |
|---|---|---|
| `internal/proxy` | 94.9% | 94.9% |
| `internal/quota` | 99.3% | 99.3% |
| `internal/failover` | 95.6% | 95.6% |
| `internal/model` | 95.5% | 95.5% |

Identical. This is a strictly cheaper way to compute the same answer, not a softer gate. During testing it still failed correctly on an uncovered changed line, with the same message as before.

## A false start worth recording

This looked unnecessary at first. Probing by adding an unused function to a package showed dependents staying cached, suggesting Go's cache already handled it. That probe was broken: an unused function is stripped by the linker, so the test binary comes out byte-identical and the cached result is legitimately still valid. It was measuring dead code being deleted.

A real change behaves differently. `internal/api` sits downstream of 24 of the 33 `internal/` packages, and editing an error string in `internal/db` is enough to re-run all ~190s of it. The cache helps, but not for the changes you actually make.

## Edge case closed

The diff lists deleted paths too, and a package whose last file was removed no longer exists on disk. `go test ./internal/gone` is a hard error, so an unrelated deletion would have blocked the push for a reason with nothing to do with coverage. Only directories still present are passed.

Falls back to the whole tree when there is no merge-base to scope by, matching the existing conservative default in `touched()`. CI is unchanged and still enforces coverage against the real ref.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/pre-push [hunk 1]
```diff
@@ -223,11 +223,51 @@ else
...
+				# for a reason that has nothing to do with coverage. Keep only
+				# directories that are still there.
+				while read -r _dir; do
+					[ -n "$_dir" ] || continue
+					[ -d "$REPO_ROOT/$_dir" ] || continue
+					GO_PKGS="$GO_PKGS ./$_dir"
+				done < <(printf '%s\n' "$CHANGED_FILES" |
+					grep -E '^internal/.*\.go$' |
+					grep -vE '_test\.go$' |
+					xargs -d '\n' -r -n1 dirname |
```
[the remark this answers is not in the document]
Comment (hugalafutro) [reply scripts/pre-push]: Correct, fixed in fa26baaf. The failure mode you describe is the annoying kind: not a broken push, just a silent loss of the entire point of the change, plus an option error in the output.

Dropped `xargs` from the package discovery altogether rather than reaching for a portable invocation, since `sed 's#/[^/]*$##'` does the job directly and is one process instead of two. Every path reaching it matched `^internal/.*\.go$`, so it always contains a slash.

While in there I fixed the same GNU-only construct in `has_measurable_change`, which predates this PR:

```diff
-diff="$(printf '%s\n' "$files" | xargs -d '\n' -r git diff -U0 "$SYNC_BASE...HEAD" -- 2>/dev/null)" || return 0
+diff="$(printf '%s\n' "$files" | tr '\n' '\0' | xargs -0 git diff -U0 "$SYNC_BASE...HEAD" -- 2>/dev/null)" || return 0
```

Same root cause with a worse consequence: on macOS that errored, hit the `|| return 0`, and made the function return true unconditionally, so every push ran the full coverage suite regardless of what changed. `-r` was also GNU-only and is unnecessary here, since the guard above already returns when `files` is empty.

Re-ran the end-to-end check after the change, same realistic diff as the PR description:

```
pre-push: coverage - go test ./internal/db (Postgres :5433 up)
Diff coverage: 1/1 = 100.0% (threshold 90%)
ELAPSED: 8s
```

Scoping intact, gate still evaluating properly. `grep -n xargs scripts/pre-push` now returns only the `-0` form and the comments explaining why.
